src/annotator/util/normalize.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * Find the smallest offset in `str` which contains at least `count` chars
+ * that match `filter` before it.
+ *
+ * @param {string} str
+ * @param {number} count
+ * @param {(char: string) => boolean} filter
+ * @param {number} [startPos]
+ */
+function advance(str, count, filter, startPos = 0) {
+  let pos = startPos;
+  while (pos < str.length && count > 0) {
+    if (filter(str[pos])) {
+      --count;
+    }
+    ++pos;
+  }
+  return pos;
+}
+
+/**
+ * Count characters which match `filter` in `str`.
+ *
+ * @param {string} str
+ * @param {(char: string) => boolean} filter
+ * @param {number} startPos
+ * @param {number} endPos
+ */
+function countChars(str, filter, startPos, endPos) {
+  let count = 0;
+  for (let pos = startPos; pos < endPos; pos++) {
+    if (filter(str[pos])) {
+      ++count;
+    }
+  }
+  return count;
+}
+
+/**
+ * Translate a (start, end) pair of offsets in an input string to corresponding
+ * offsets in an output string, where positions in the input and output strings
+ * are related by counting the number of preceding characters that pass a
+ * filter.
+ *
+ * In typical usage, the output string is a corrupted version of the input
+ * string, where characters that do not match the filter (eg. whitespace)
+ * have been inserted or deleted at arbitrary locations.
```

Comment:
The terms input/output originate from before this was extracted into a generic utility. Perhaps it would be better to just call them "stringA" and "stringB" or something. The important idea is that stringB is a copy of stringA which has been modified by adding or removing whitespace (or some other characters determined by `filter`) at arbitrary locations. The reason that `filter` is an argument to this function is because different callers might have different ideas of what spaces (or other characters) may have been added or removed by the corruption process.

Reply:
I think `stringA` and `stringB` might be better. Is it more accurate to say "`stringB` is a copy of `stringA`" or that "`stringA` and `stringB` are strings representing the same textual content with potential differences in whitespace"? 

I also think the word "corrupted" in the function docstring is throwing me.